### PR TITLE
refactor: check assertions per import statement

### LINKED ISSUE #132: Support multiple ranges for an import in the graph
Currently the graph cannot represent multiple locations an import specifier might be imported in a file.  For example, an import can be like this:

```ts
import * as a from "https://example.com/a";
import { default as b } from "https://example.com/a";
```

In addition, in theory, a module can be statically and dynamically imported:

```ts
import * as a from "https://example.com/a";
const b = await import("https://example.com/a";
```

Or in TypeScript the same specifier can be provided for a type only import:

```ts
import { a } from "https://example.com/a";
import type { A } from "https://example.com/a";
```

Currently we only store the last range for the specifier in the graph.  This makes it difficult when trying to rewrite specifiers (like in dnt or `deno vendor`) as well as means that diagnostics are only issued for the "last" import location, not all import locations for a given specifier.

We should enhance the graph to be able to store this information, which is likely going to be a breaking change, but one that increases the usability of the graph.

### LINKED ISSUE #160: Don't enforce assertions on type-only imports
There exists `TS2822 [ERROR]: Import assertions cannot be used with type-only imports or exports.`, so we shouldn't check assertions for type-only imports. https://github.com/denoland/deno/issues/14913

### PR DESCRIPTION
**Pull request by nayeemrmn**
Closes #132
Fixes #160
Supersedes #155

Assertions appropriately no longer influence module loading/parsing. Instead they are checked in a pass at the end; resulting errors are stored in `Import::errors: Vec<ImportError>`. 'Import errors' are a third kind of error alongside module slot errors and resolution errors which are now checked in `ModuleEntryIterator::validate()`.

### CODE REVIEW

File: src/analyzer.rs
Comment (dsherret) [inline src/analyzer.rs:190]: Can you change `_` here to be explicit about the variant? That will make the code easier to read without knowing all the variants and cause compile errors here when the variants change. I know we don't do that everywhere in the code here, but we should do it more often.

File: src/analyzer.rs
Comment (dsherret) [inline src/analyzer.rs:186]: I'm confused by this method name because it does `!map.contains_key`. Maybe this should be `has(&self, key: &str)` then we can invert everything and do `!has(` when using it?

File: src/analyzer.rs
Comment (dsherret) [inline src/analyzer.rs:213]: We should convert this wildcard to be explicit.

File: src/graph.rs
Comment (nayeemrmn) [inline src/graph.rs:1747]: This is the current method for handling different assertions being made against the same module... re-parse it with each one (because that's where assertion errors are generated). And it seems like there's no uniqueness in the assert types in `pending_specifiers` because it includes the range... so are we re-parsing the module for every import that was found for it until it was visited?

File: src/analyzer.rs
Comment (dsherret) [inline src/analyzer.rs:199]: Where is this used? Is it necessary? We should minimize having methods on the that allocate while analyzing a graph. For example, maybe this should instead be a method that verifies the `ImportAssertions` then someone could use `.get` afterwards instead or this should return a new struct where this method verifies the `ImportAssertions` then has a reference to the old map while providing some `.get` methods.

File: src/analyzer.rs
Comment (dsherret) [inline src/analyzer.rs:200]: If really necessary, move within the match and specify the capacity based on the existing map:

```suggestion
    let mut assertions = HashMap::with_capacity(map.len());
```

File: src/analyzer.rs
Comment (dsherret) [inline src/analyzer.rs:209]: We do we return `None` for everything if a specific import assertion is not known? Maybe this method seems too specific and shouldn't exist in deno_graph?

File: src/graph.rs
Comment (dsherret) [inline src/graph.rs:160]: We should remove this panic. Either all the errors should have a specifier or this should become `Option<&ModuleSpecifier>`. I'm not sure what's changed that there wouldn't be a specifier anymore though?

**Review by dsherret — COMMENTED** [review]
I'll review the rest of it later.

File: src/graph.rs
Comment (dsherret) [inline src/graph.rs:1752]: I'm not sure we need this because this information should be avaiable straight from the graph? I think this will maybe fail if someone calls `build` multiple times.

File: src/analyzer.rs
Comment (nayeemrmn) [reply src/analyzer.rs:199]: Ah I was going to use this as an alternative serialization for static imports but changed my mind (deferred for the future?), I'll remove this method.

File: src/graph.rs
Comment (nayeemrmn) [reply src/graph.rs:160]: > I'm not sure what's changed that there wouldn't be a specifier anymore though?

Yeah I've been meaning to complain about this method... the specifier returned by `ResolutionError` has a drastically different meaning than the other variants (referrer vs target). Similar with `maybe_range()`. `ImportError` kind of aligns with `ResolutionError` on that.

It's a symptom of `ModuleGraphError` mixing errors that can occupy module slots vs errors that represent arrows between them. A year ago on a previous attempt at this fix I tried separating them more thoroughly but the boilerplate was extreme. So yeah I think we should just make it return an `Option`.

File: src/graph.rs
Comment (nayeemrmn) [reply src/graph.rs:1751]: It's to work around borrow issues at https://github.com/denoland/deno_graph/blob/af1009eb0e5cd9dcb1c851911f48268ed187f7b5/src/graph.rs#L1877-L1885
If there were a workaround limited to `Builder::fill()`, it would probably be to build up the same map but as its own pass just above this.

File: src/analyzer.rs
Comment (nayeemrmn) [reply src/analyzer.rs:186]: Then `has()` would have to return true if the presence of associated value of that key is unknown. The method returns true if the given key is 100% known to be absent (that's where we treat it the assertion as `type: "js"`), and returns false if the set of keys is dynamic (can't check assertions) or the key is set to a value (we will use `get()` to check assertions).

File: src/graph.rs
Comment (dsherret) [reply src/graph.rs:1751]: Yeah, we'll have to make that work based on the current module graph. It would be better if we didn't create two sources of truth and just derived from the graph.

File: src/analyzer.rs
Comment (dsherret) [reply src/analyzer.rs:186]: True. Maybe add a documentation comment on this?

File: src/graph.rs
Comment (dsherret) [inline src/graph.rs:149]: RE https://github.com/denoland/deno_graph/pull/228#discussion_r1103865391 -- I wonder if we should have `ModuleError(ModuleError)` then put `LoadingErr`, `Missing`, `ParseError`, and `UnsupportedMediaType` in there. Then we could move `pub fn specifier(&self) -> &ModuleSpecifier` onto that and then move `maybe_range` into the other errors. What do you think?

File: src/graph.rs
Comment (dsherret) [inline src/graph.rs:557]: I'm not sure if "imports" is the right word here because currently in this PR this could be imports, exports, or typescript references. Perhaps it should be `specifiers` instead?

File: src/graph.rs
Comment (dsherret) [inline src/graph.rs:1651]: Is this done later? Why aren't dynamic imports checked? Maybe include the answer to "why?" in the comment here.

**Review by dsherret — CHANGES_REQUESTED** [review]
Thanks for this @nayeemrmn. It's looking better than main, but I'm not sure about the name "imports". Perhaps this should be something more generic because it also includes exports and typescript path references?

File: src/graph.rs
Comment (nayeemrmn) [reply src/graph.rs:1651]: The way we discover dynamic module slot errors in `deno_graph` and defer elevating them until runtime, there's no good mechanism for doing that currently in CLI for resolution/import errors so we just leave it for `deno_core`. That and there was a test for it (`test_build_graph_dynamic_json_ignores_assert`) so I just excluded it for simplicity.

I think I'll move this check to `validate()` actually, it's more consistent with module slot errors.

Comment (dsherret) [reply src/graph.rs:1651]: Good idea. That sounds a lot better.

File: src/graph.rs
Comment (nayeemrmn) [reply src/graph.rs:149]: Yeah that's what I did in the previous attempt from last year I mentioned https://github.com/nayeemrmn/deno_graph/blob/786128a3f6330b1bdf7c6408e2ca53eb9c4688e9/src/graph.rs#L297-L302. I'll try it again, it should go better with this version's solution.

File: test.ts
Comment (nayeemrmn) [reply test.ts:303]: Yeah, I guess we can switch to blessed fields for known assertions on the serialization layer. Just now wondering if we should store it like that in the first place..

File: src/graph.rs
Comment (dsherret) [reply src/graph.rs:557]: Ok, imports doesn't sound too bad. You're right that it's more clear.

File: test.ts
Comment (dsherret) [reply test.ts:303]: If everything still works then storing it like that might be better.

Comment (nayeemrmn) [reply test.ts:303]: I've been mulling over schemas that:
1. Transfer well to dynamic imports with potential unknowns
2. Don't then add a ton of boilerplate for static imports
3. Don't conflate dynamic/unknown assertions with unspecified ones

And landed on:
```ts
interface Import {
  assertions: "unknown" | Record<string, string>,
}
```
where for static imports you can just pretend the `"unknown" |` isn't there. This would use the `maybe_static()` method from https://github.com/denoland/deno_graph/pull/228#discussion_r1103861394 and just mark dynamic imports as `"unknown"` if there is _any_ dynamic part of it. We have to give up the blessed fields like `assertionType?: string` since it would violate either (1) or (3) but `assertions.type` is pretty much the same.

File: src/graph.rs
Comment (dsherret) [inline src/graph.rs:886]: I think this should be:

```suggestion
                if self.follow_dynamic || !import.is_dynamic {
```

I just realized the above on line 872 is wrong and should also be similar.

Comment (nayeemrmn) [reply src/graph.rs]: Note to correct https://github.com/denoland/deno/blob/f917d2e2c10e0a94e564a9016217e7ce27c8bbee/cli/tools/vendor/build.rs#L88 to false

Comment (dsherret) [reply src/graph.rs]: I changed that one to `true` because I think we should validate dynamic imports for vendoring.

File: test.ts
Comment (dsherret) [reply test.ts:303]: Sounds good.

File: src/graph.rs
Comment (nayeemrmn) [reply src/graph.rs]: Okay I thought this had been masking errors that would now be exposed but it only applies to resolution errors. I can see why missing modules should be caught in advance for vendoring

Comment (dsherret) [reply src/graph.rs]: That said, I guess that would prevent vendoring of the scenario you described yesterday where someone might conditionally load some plugin code or something. Maybe instead it should be warnings.

File: src/graph.rs
Comment (nayeemrmn) [inline src/graph.rs:566]: I've merged `ModuleKind::Asserted` into `ModuleKind::Esm`. That info is given by `Module::media_type` and assert-worthy module types should fall in naturally for other purposes. The name seems more related to module system not language so it should be fine.

Comment (dsherret) [reply src/graph.rs:566]: I'm doing a large refactor for npm specifiers atm that's going to remove `ModuleKind` and replace it with an `enum Module`. We're not going to land this PR until deno 2.0 so you'll get some conflicts on this on this one, but I don't think it will be too much trouble.

File: src/graph.rs
Comment (dsherret) [inline src/graph.rs:141]: By the way, we can't land this PR until probably next month in preparation for Deno 2.0. If you want to get some of these non-breaking changes into main faster then you might want to consider refactoring these out ot a separate PR and I'll take a look at it to get it merged quickly.

File: src/graph.rs
Comment (dsherret) [reply src/graph.rs:566]: I think we could represent it as `Module::Json` instead of `Module::Asserted`. I'll update my PR to that.

**Comment by nayeemrmn** [comment]
Regarding 2.0 breakages, seems like the only breakage is `Dependency::assertionType`. Could we keep that as legacy for a month to unblock this PR?

**Comment by dsherret** [comment]
Probably not because merging this would also require more work upgrading the rest of the code in other repos and that would block progress on getting npm specifiers in deno_graph, which is the priority at the moment because we want to get them in a more ideal state for integration with Deploy and other tooling. It's just waiting a couple weeks until early March.

File: src/graph.rs
Comment (dsherret) [inline src/graph.rs:149]: Would you be able to pull out these error changes and open a separate PR with them (and for anything else in this PR that's non-breaking for the json)? I'm not sure when we'll be able to land this PR and it would be nice to get this change in.

**Comment by nayeemrmn** [comment]
@dsherret I went ahead and preserved `assertionType` in the serialization for legacy. It's a very self-contained, small addition in `serialize_dependencies()`. This can be considered for landing without breaking `deno info` now.

**Comment by nayeemrmn** [comment]
With import assertions becoming import attributes and them being part of the cache key, this PR goes in the wrong direction for the future. Closing.

**Pull request closed without merging**
